lib/bureaucrat/helpers.ex

```diff
@@ -125,24 +125,58 @@ defmodule Bureaucrat.Helpers do
       opts =
         opts
         |> Keyword.put_new(:operation_id, default_operation_id)
+        |> Keyword.update!(:description, &(&1 || default_description()))
+
+      if Keyword.fetch!(opts, :description) != nil do
+        Bureaucrat.Recorder.doc(conn, opts)
+      else
+        file = Keyword.fetch!(opts, :file)
+        line = Keyword.fetch!(opts, :line)
+
+        Mix.shell().info("""
+        The request at #{file}:#{line} won't be recorded by bureaucrat because
+        the description can't be determined and none is explicitly provided.
+        To address this, you can pass the :description option to this macro.
+
+        If this macro is invoked indirectly, via the request macros, such as
+        get and post, you can switch to the _undocumented version to
+        explicitly avoid generating the documentation.
+
+        Alternatively, you can provide the description manually with something
+        like doc(conn, description: "some description"), where conn is the result
+        of the _undocumented macro.
+        """)
+      end
 
-      Bureaucrat.Recorder.doc(conn, opts)
       conn
     end
   end
 
-  def format_test_name("test " <> name), do: name
+  def default_description do
+    # We look for the first test process in the stack traces of this process as well as other callers.
+    # The latter allows us to find the test description even if this function is running in a task started by a test process.
+    callers = Process.get(:"$callers") || []
+    Enum.find_value([self() | callers], &test_description/1)
+  end
 
-  def format_test_name(function_name) do
-    raise """
-    It looks like you called a `Phoenix.ConnTest` macro inside `#{function_name}`.
-    Bureaucrat can only document macros `get`, `post`, `delete`, etc. when they are called inside a `test` block.
+  defp test_description(pid) do
+    {:current_stacktrace, stacktrace} = Process.info(pid, :current_stacktrace)
 
-    If the request macro is called inside a private function or setup, you should explicitly say you don't want Bureaucrat to document this request.
-    Use `get_undocumented`, `post_undocumented`, `delete_undocumented`, `patch_undocumented` or `put_undocumented` instead.
-    """
+    Enum.find_value(
+      stacktrace,
+      fn {module, function, _arity, opts} ->
+        with "test " <> description <- to_string(function),
+             true <- String.ends_with?(to_string(module), "Test"),
+             true <- String.ends_with?(to_string(Keyword.get(opts, :file)), ".exs"),
+             do: description,
+             else: (_ -> nil)
+      end
+    )
   end
 
+  def format_test_name("test " <> name), do: name
+  def format_test_name(_other), do: nil
+
```

Comment:
this is not a review of your code, but i don't like how git recorded this diff (or how GH rendered it?? this makes me realize i don't exactly know how diffs are stored)

Reply:
I think it looks more intuitive if you're reading one commit at a time, because then the new empty line is added before `format_test_name`, not after.